Application crashes if the chain is stopped

**Describe the bug**

The tui crashes if you stop the chain that is listening to.

**To Reproduce**
Steps to reproduce the behavior:

* Start a chain in a terminal windown e.g. `$ gaiad start`
* On another terminal run the tool `$ gex`
* Go back to the terminal that was running the chain and stop it
* The tui will show the error below

```
panic: Get "http://localhost:26657/heal": dial tcp [::1]:26657: connect: connection refused
go routine 55 [running]:
...
```

**Expected behavior**

Probably in the routine listening on the port, some error handling could gracefully exit the tool or if possible (maybe more complex), show an error in the `Health` widget.

**Screenshot**

![gex_error](https://user-images.githubusercontent.com/45477427/100124896-31a0a000-2e4a-11eb-865d-64864c81fd51.png)

--- a/LICENSE
+++ b/LICENSE
@@ -1,0 +1,203 @@
+License: Apache2.0
+
+                                 Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
+
+   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
+
+   1. Definitions.
+
+      "License" shall mean the terms and conditions for use, reproduction,
+      and distribution as defined by Sections 1 through 9 of this document.
+
+      "Licensor" shall mean the copyright owner or entity authorized by
+      the copyright owner that is granting the License.
+
+      "Legal Entity" shall mean the union of the acting entity and all
+      other entities that control, are controlled by, or are under common
+      control with that entity. For the purposes of this definition,
+      "control" means (i) the power, direct or indirect, to cause the
+      direction or management of such entity, whether by contract or
+      otherwise, or (ii) ownership of fifty percent (50%) or more of the
+      outstanding shares, or (iii) beneficial ownership of such entity.
+
+      "You" (or "Your") shall mean an individual or Legal Entity
+      exercising permissions granted by this License.
+
+      "Source" form shall mean the preferred form for making modifications,
+      including but not limited to software source code, documentation
+      source, and configuration files.
+
+      "Object" form shall mean any form resulting from mechanical
+      transformation or translation of a Source form, including but
+      not limited to compiled object code, generated documentation,
+      and conversions to other media types.
+
+      "Work" shall mean the work of authorship, whether in Source or
+      Object form, made available under the License, as indicated by a
+      copyright notice that is included in or attached to the work
+      (an example is provided in the Appendix below).
+
+      "Derivative Works" shall mean any work, whether in Source or Object
+      form, that is based on (or derived from) the Work and for which the
+      editorial revisions, annotations, elaborations, or other modifications
+      represent, as a whole, an original work of authorship. For the purposes
+      of this License, Derivative Works shall not include works that remain
+      separable from, or merely link (or bind by name) to the interfaces of,
+      the Work and Derivative Works thereof.
+
+      "Contribution" shall mean any work of authorship, including
+      the original version of the Work and any modifications or additions
+      to that Work or Derivative Works thereof, that is intentionally
+      submitted to Licensor for inclusion in the Work by the copyright owner
+      or by an individual or Legal Entity authorized to submit on behalf of
+      the copyright owner. For the purposes of this definition, "submitted"
+      means any form of electronic, verbal, or written communication sent
+      to the Licensor or its representatives, including but not limited to
+      communication on electronic mailing lists, source code control systems,
+      and issue tracking systems that are managed by, or on behalf of, the
+      Licensor for the purpose of discussing and improving the Work, but
+      excluding communication that is conspicuously marked or otherwise
+      designated in writing by the copyright owner as "Not a Contribution."
+
+      "Contributor" shall mean Licensor and any individual or Legal Entity
+      on behalf of whom a Contribution has been received by Licensor and
+      subsequently incorporated within the Work.
+
+   2. Grant of Copyright License. Subject to the terms and conditions of
+      this License, each Contributor hereby grants to You a perpetual,
+      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
+      copyright license to reproduce, prepare Derivative Works of,
+      publicly display, publicly perform, sublicense, and distribute the
+      Work and such Derivative Works in Source or Object form.
+
+   3. Grant of Patent License. Subject to the terms and conditions of
+      this License, each Contributor hereby grants to You a perpetual,
+      worldwide, non-exclusive, no-charge, royalty-free, irrevocable
+      (except as stated in this section) patent license to make, have made,
+      use, offer to sell, sell, import, and otherwise transfer the Work,
+      where such license applies only to those patent claims licensable
+      by such Contributor that are necessarily infringed by their
+      Contribution(s) alone or by combination of their Contribution(s)
+      with the Work to which such Contribution(s) was submitted. If You
+      institute patent litigation against any entity (including a
+      cross-claim or counterclaim in a lawsuit) alleging that the Work
+      or a Contribution incorporated within the Work constitutes direct
+      or contributory patent infringement, then any patent licenses
+      granted to You under this License for that Work shall terminate
+      as of the date such litigation is filed.
+
+   4. Redistribution. You may reproduce and distribute copies of the
+      Work or Derivative Works thereof in any medium, with or without
+      modifications, and in Source or Object form, provided that You
+      meet the following conditions:
+
+      (a) You must give any other recipients of the Work or
+          Derivative Works a copy of this License; and
+
+      (b) You must cause any modified files to carry prominent notices
+          stating that You changed the files; and
+
+      (c) You must retain, in the Source form of any Derivative Works
+          that You distribute, all copyright, patent, trademark, and
+          attribution notices from the Source form of the Work,
+          excluding those notices that do not pertain to any part of
+          the Derivative Works; and
+
+      (d) If the Work includes a "NOTICE" text file as part of its
+          distribution, then any Derivative Works that You distribute must
+          include a readable copy of the attribution notices contained
+          within such NOTICE file, excluding those notices that do not
+          pertain to any part of the Derivative Works, in at least one
+          of the following places: within a NOTICE text file distributed
+          as part of the Derivative Works; within the Source form or
+          documentation, if provided along with the Derivative Works; or,
+          within a display generated by the Derivative Works, if and
+          wherever such third-party notices normally appear. The contents
+          of the NOTICE file are for informational purposes only and
+          do not modify the License. You may add Your own attribution
+          notices within Derivative Works that You distribute, alongside
+          or as an addendum to the NOTICE text from the Work, provided
+          that such additional attribution notices cannot be construed
+          as modifying the License.
+
+      You may add Your own copyright statement to Your modifications and
+      may provide additional or different license terms and conditions
+      for use, reproduction, or distribution of Your modifications, or
+      for any such Derivative Works as a whole, provided Your use,
+      reproduction, and distribution of the Work otherwise complies with
+      the conditions stated in this License.
+
+   5. Submission of Contributions. Unless You explicitly state otherwise,
+      any Contribution intentionally submitted for inclusion in the Work
+      by You to the Licensor shall be under the terms and conditions of
+      this License, without any additional terms or conditions.
+      Notwithstanding the above, nothing herein shall supersede or modify
+      the terms of any separate license agreement you may have executed
+      with Licensor regarding such Contributions.
+
+   6. Trademarks. This License does not grant permission to use the trade
+      names, trademarks, service marks, or product names of the Licensor,
+      except as required for reasonable and customary use in describing the
+      origin of the Work and reproducing the content of the NOTICE file.
+
+   7. Disclaimer of Warranty. Unless required by applicable law or
+      agreed to in writing, Licensor provides the Work (and each
+      Contributor provides its Contributions) on an "AS IS" BASIS,
+      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+      implied, including, without limitation, any warranties or conditions
+      of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
+      PARTICULAR PURPOSE. You are solely responsible for determining the
+      appropriateness of using or redistributing the Work and assume any
+      risks associated with Your exercise of permissions under this License.
+
+   8. Limitation of Liability. In no event and under no legal theory,
+      whether in tort (including negligence), contract, or otherwise,
+      unless required by applicable law (such as deliberate and grossly
+      negligent acts) or agreed to in writing, shall any Contributor be
+      liable to You for damages, including any direct, indirect, special,
+      incidental, or consequential damages of any character arising as a
+      result of this License or out of the use or inability to use the
+      Work (including but not limited to damages for loss of goodwill,
+      work stoppage, computer failure or malfunction, or any and all
+      other commercial damages or losses), even if such Contributor
+      has been advised of the possibility of such damages.
+
+   9. Accepting Warranty or Additional Liability. While redistributing
+      the Work or Derivative Works thereof, You may choose to offer,
+      and charge a fee for, acceptance of support, warranty, indemnity,
+      or other liability obligations and/or rights consistent with this
+      License. However, in accepting such obligations, You may act only
+      on Your own behalf and on Your sole responsibility, not on behalf
+      of any other Contributor, and only if You agree to indemnify,
+      defend, and hold each Contributor harmless for any liability
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+
+   END OF TERMS AND CONDITIONS
+
+   APPENDIX: How to apply the Apache License to your work.
+
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "{}"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2016 All in Bits, Inc
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.

--- a/go.mod
+++ b/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/OzqurYalcin/google-analytics v0.0.0-20201022181344-721c3dab31b5
 	github.com/go-openapi/strfmt v0.19.5 // indirect
 	github.com/google/uuid v1.1.1
-	github.com/gorilla/websocket v1.4.2 // indirect
+	github.com/gorilla/websocket v1.4.2
 	github.com/jedib0t/go-pretty v4.3.0+incompatible
 	github.com/mum4k/termdash v0.12.2
 	github.com/sacOO7/go-logger v0.0.0-20180719173527-9ac9add5a50d // indirect

--- a/main.go
+++ b/main.go
@@ -328,6 +328,8 @@ func writeHealth(ctx context.Context, t *text.Text, delay time.Duration, connect
 				t.Write("🔴 no connection")
 				if reconnect == false {
 					connectionSignal <- "no_connection"
+					connectionSignal <- "no_connection"
+					connectionSignal <- "no_connection"
 					reconnect = true
 				}
 			}

--- a/main.go
+++ b/main.go
@@ -14,7 +14,7 @@
 // limitations under the License.
 
 // Binary explorer demo. Displays widgets for insights of blockchain behaviour.
-// Exist when 'q' is pressed.
+// Exist when 'q' or 'esc' is pressed.
 package main
 
 import (
@@ -24,8 +24,6 @@ import (
 	"time"
 
 	"log"
-	"os"
-	"os/signal"
 
 	"gopkg.in/resty.v1"
 
@@ -55,6 +53,7 @@ var givenPort = flag.String("p", "26657", "port to connect to as a string")
 
 func main() {
 	view()
+	connectionSignal := make(chan string)
 	t, err := termbox.New()
 	if err != nil {
 		panic(err)
@@ -65,6 +64,9 @@ func main() {
 
 	networkInfo := getFromRPC("status")
 	networkStatus := gjson.Parse(networkInfo)
+	if !networkStatus.Exists() {
+		panic("Application not running on localhost:" + fmt.Sprintf("%s", *givenPort))
+	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 
@@ -131,14 +133,14 @@ func main() {
 	// system powered widgets
 	go writeTime(ctx, timeWidget, 1*time.Second)
 
-	// websocket powered widgets
-	go writeValidators(ctx, validatorWidget)
-	go writeBlocks(ctx, blocksWidget)
-	go writeTransactions(ctx, transactionWidget)
-
 	// rpc widgets
 	go writePeers(ctx, peerWidget, 1*time.Second)
-	go writeHealth(ctx, healthWidget, 1*time.Second)
+	go writeHealth(ctx, healthWidget, 500*time.Millisecond, connectionSignal)
+
+	// websocket powered widgets
+	go writeValidators(ctx, validatorWidget, connectionSignal)
+	go writeBlocks(ctx, blocksWidget, connectionSignal)
+	go writeTransactions(ctx, transactionWidget, connectionSignal)
 
 	// blockchain download gauge
 	syncWidget, err := gauge.New(
@@ -253,14 +255,10 @@ func main() {
 
 func getFromRPC(endpoint string) string {
 	port := *givenPort
-	resp, err := resty.R().
+	resp, _ := resty.R().
 		SetHeader("Cache-Control", "no-cache").
 		SetHeader("Content-Type", "application/json").
 		Get(appRPC + ":" + port + "/" + endpoint)
-
-	if err != nil {
-		panic(err)
-	}
 
 	return resp.String()
 }
@@ -300,17 +298,14 @@ func writeTime(ctx context.Context, t *text.Text, delay time.Duration) {
 
 // writeHealth writes the status to the healthWidget.
 // Exits when the context expires.
-func writeHealth(ctx context.Context, t *text.Text, delay time.Duration) {
+func writeHealth(ctx context.Context, t *text.Text, delay time.Duration, connectionSignal chan string) {
+	reconnect := false
 	health := gjson.Get(getFromRPC("health"), "result")
 	t.Reset()
 	if health.Exists() {
-		if err := t.Write("🟢 good"); err != nil {
-			panic(err)
-		}
+		t.Write("🟢 good")
 	} else {
-		if err := t.Write("🔴 no connection"); err != nil {
-			panic(err)
-		}
+		t.Write("🔴 no connection")
 	}
 
 	ticker := time.NewTicker(delay)
@@ -322,15 +317,20 @@ func writeHealth(ctx context.Context, t *text.Text, delay time.Duration) {
 			health := gjson.Get(getFromRPC("health"), "result")
 			t.Reset()
 			if health.Exists() {
-				if err := t.Write("🟢 good"); err != nil {
-					panic(err)
+				t.Write("🟢 good")
+				if reconnect == true {
+					connectionSignal <- "reconnect"
+					connectionSignal <- "reconnect"
+					connectionSignal <- "reconnect"
+					reconnect = false
 				}
 			} else {
-				if err := t.Write("🔴 no connection"); err != nil {
-					panic(err)
+				t.Write("🔴 no connection")
+				if reconnect == false {
+					connectionSignal <- "no_connection"
+					reconnect = true
 				}
 			}
-
 		case <-ctx.Done():
 			return
 		}
@@ -342,6 +342,9 @@ func writeHealth(ctx context.Context, t *text.Text, delay time.Duration) {
 func writePeers(ctx context.Context, t *text.Text, delay time.Duration) {
 	peers := gjson.Get(getFromRPC("net_info"), "result.n_peers").String()
 	t.Reset()
+	if peers != "" {
+		t.Write(peers)
+	}
 	if err := t.Write(peers); err != nil {
 		panic(err)
 	}
@@ -354,8 +357,8 @@ func writePeers(ctx context.Context, t *text.Text, delay time.Duration) {
 		case <-ticker.C:
 			t.Reset()
 			peers := gjson.Get(getFromRPC("net_info"), "result.n_peers").String()
-			if err := t.Write(peers); err != nil {
-				panic(err)
+			if peers != "" {
+				t.Write(peers)
 			}
 
 		case <-ctx.Done():
@@ -366,16 +369,10 @@ func writePeers(ctx context.Context, t *text.Text, delay time.Duration) {
 
 // writeTransactions writes the latest Transactions to the transactionsWidget.
 // Exits when the context expires.
-func writeTransactions(ctx context.Context, t *text.Text) {
-	interrupt := make(chan os.Signal, 1)
-	signal.Notify(interrupt, os.Interrupt)
-
+func writeTransactions(ctx context.Context, t *text.Text, connectionSignal <-chan string) {
 	port := *givenPort
 	socket := gowebsocket.New("ws://localhost:" + port + "/websocket")
 
-	socket.OnConnectError = func(err error, socket gowebsocket.Socket) {
-		log.Fatal("Received connect error - ", err)
-	}
 	socket.OnTextMessage = func(message string, socket gowebsocket.Socket) {
 		currentTx := gjson.Get(message, "result.data.value.TxResult.result.log")
 		currentTime := time.Now()
@@ -386,17 +383,19 @@ func writeTransactions(ctx context.Context, t *text.Text) {
 		}
 	}
 
-	socket.OnDisconnected = func(err error, socket gowebsocket.Socket) {
-		log.Println("Disconnected from server ")
-		return
-	}
-
 	socket.Connect()
 
 	socket.SendText("{ \"jsonrpc\": \"2.0\", \"method\": \"subscribe\", \"params\": [\"tm.event='Tx'\"], \"id\": 2 }")
 
 	for {
 		select {
+		case s := <-connectionSignal:
+			if s == "no_connection" {
+				socket.Close()
+			}
+			if s == "reconnect" {
+				writeTransactions(ctx, t, connectionSignal)
+			}
 		case <-ctx.Done():
 			log.Println("interrupt")
 			socket.Close()
@@ -407,16 +406,10 @@ func writeTransactions(ctx context.Context, t *text.Text) {
 
 // writeBlocks writes the latest Block to the blocksWidget.
 // Exits when the context expires.
-func writeBlocks(ctx context.Context, t *text.Text) {
-	interrupt := make(chan os.Signal, 1)
-	signal.Notify(interrupt, os.Interrupt)
+func writeBlocks(ctx context.Context, t *text.Text, connectionSignal <-chan string) {
 
 	port := *givenPort
 	socket := gowebsocket.New("ws://localhost:" + port + "/websocket")
-
-	socket.OnConnectError = func(err error, socket gowebsocket.Socket) {
-		log.Fatal("Received connect error - ", err)
-	}
 
 	socket.OnTextMessage = func(message string, socket gowebsocket.Socket) {
 		currentBlock := gjson.Get(message, "result.data.value.block.header.height")
@@ -428,17 +421,19 @@ func writeBlocks(ctx context.Context, t *text.Text) {
 
 	}
 
-	socket.OnDisconnected = func(err error, socket gowebsocket.Socket) {
-		log.Println("Disconnected from server ")
-		return
-	}
-
 	socket.Connect()
 
 	socket.SendText("{ \"jsonrpc\": \"2.0\", \"method\": \"subscribe\", \"params\": [\"tm.event='NewBlock'\"], \"id\": 1 }")
 
 	for {
 		select {
+		case s := <-connectionSignal:
+			if s == "no_connection" {
+				socket.Close()
+			}
+			if s == "reconnect" {
+				writeBlocks(ctx, t, connectionSignal)
+			}
 		case <-ctx.Done():
 			log.Println("interrupt")
 			socket.Close()
@@ -449,16 +444,9 @@ func writeBlocks(ctx context.Context, t *text.Text) {
 
 // writeValidators writes the current validator set to the validatoWidget
 // Exits when the context expires.
-func writeValidators(ctx context.Context, t *text.Text) {
-	interrupt := make(chan os.Signal, 1)
-	signal.Notify(interrupt, os.Interrupt)
-
+func writeValidators(ctx context.Context, t *text.Text, connectionSignal <-chan string) {
 	port := *givenPort
 	socket := gowebsocket.New("ws://localhost:" + port + "/websocket")
-
-	socket.OnConnectError = func(err error, socket gowebsocket.Socket) {
-		log.Fatal("Received connect error - ", err)
-	}
 
 	socket.OnConnected = func(socket gowebsocket.Socket) {
 		validators := gjson.Get(getFromRPC("validators"), "result.validators")
@@ -495,17 +483,19 @@ func writeValidators(ctx context.Context, t *text.Text) {
 		})
 	}
 
-	socket.OnDisconnected = func(err error, socket gowebsocket.Socket) {
-		log.Println("Disconnected from server ")
-		return
-	}
-
 	socket.Connect()
 
 	socket.SendText("{ \"jsonrpc\": \"2.0\", \"method\": \"subscribe\", \"params\": [\"tm.event='ValidatorSetUpdates'\"], \"id\": 3 }")
 
 	for {
 		select {
+		case s := <-connectionSignal:
+			if s == "no_connection" {
+				socket.Close()
+			}
+			if s == "reconnect" {
+				writeValidators(ctx, t, connectionSignal)
+			}
 		case <-ctx.Done():
 			log.Println("interrupt")
 			socket.Close()

--- a/main.go
+++ b/main.go
@@ -315,7 +315,6 @@ func writeHealth(ctx context.Context, t *text.Text, delay time.Duration, connect
 		select {
 		case <-ticker.C:
 			health := gjson.Get(getFromRPC("health"), "result")
-			t.Reset()
 			if health.Exists() {
 				t.Write("🟢 good")
 				if reconnect == true {

